src/main/scala/com/mwz/sonar/scala/junit/JUnitReportParser.scala

```diff
@@ -0,0 +1,116 @@
+ */
+package com.mwz.sonar.scala
+package junit
+
+import java.io.File
+import java.nio.file.Path
+
+import com.mwz.sonar.scala.util.Log
+import org.sonar.api.batch.ScannerSide
+import org.sonar.api.batch.fs.{FilePredicate, FileSystem, InputFile}
+
+import scala.collection.JavaConverters._
+import scala.xml.{Elem, XML}
+
+trait JUnitReportParserAPI {
+  def parse(tests: List[Path], directories: List[File]): Map[InputFile, JUnitReport]
+}
+
+@ScannerSide
+final class JUnitReportParser(fileSystem: FileSystem) extends JUnitReportParserAPI {
+  private[this] val log = Log(classOf[JUnitReportParser], "junit")
+
+  def parse(tests: List[Path], directories: List[File]): Map[InputFile, JUnitReport] = {
+    // Get report files - xml files starting with "TEST-".
+    val reports: List[File] = reportFiles(directories)
+
+    // Parse report files.
+    val unitTestReports: List[JUnitReport] = parseReportFiles(reports)
+
+    log.debug("Unit test reports:")
+    log.debug(unitTestReports.mkString(", "))
+
+    // Convert package names into files.
+    resolveFiles(tests, unitTestReports)
+  }
+
+  /**
+   * Get report files - xml files starting with "TEST-"
+   */
+  private[junit] def reportFiles(directories: List[File]): List[File] = {
+    val reportFiles: List[File] = directories.filter(_.isDirectory).flatMap { dir =>
+      dir.listFiles((_, name) => name.startsWith("TEST-") && name.endsWith(".xml"))
+    }
+
+    if (directories.isEmpty)
+      log.error(s"The paths ${directories.mkString(", ")} are not valid directories.")
+    else if (reportFiles.isEmpty)
+      log.error(s"No report files found in ${directories.mkString(", ")}.")
+
+    reportFiles
+  }
+
+  /**
+   * Parse report files.
+   */
+  private[junit] def parseReportFiles(reports: List[File]): List[JUnitReport] = {
+    reports.map { file =>
+      val xml: Elem = XML.loadFile(file)
+      JUnitReport(
+        name = xml \@ "name",
+        tests = (xml \@ "tests").toInt,
+        errors = (xml \@ "errors").toInt,
+        failures = (xml \@ "failures").toInt,
+        skipped = (xml \@ "skipped").toInt,
+        time = (xml \@ "time").toFloat
+      )
+    }
+  }
+
+  /**
+   * Convert package names into files.
+   */
+  private[junit] def resolveFiles(
+    tests: List[Path],
+    reports: List[JUnitReport]
+  ): Map[InputFile, JUnitReport] = {
+    reports
+      .groupBy(_.name)
+      .map {
+        case (name, reports) =>
+          val filePath: String = name.replace(".", "/")
+          val predicates: List[FilePredicate] =
+            tests.map { path =>
+              fileSystem.predicates.hasPath(path.resolve(s"$filePath.scala").toString)
+            }
+          val files: Iterable[InputFile] =
+            fileSystem
+              .inputFiles(
+                fileSystem.predicates.or(predicates.asJava)
+              )
+              .asScala
+
+          // Collect the first file here.
+          files.headOption.flatMap(file => reports.headOption.map((file, _)))
```

Comment:
Given you only care for the first file... - maybe, because it only makes sense that it matches just one file
Shouldn't you log a warning if more than one file is matched and also if any file is matched?
BTW, this reminds me #64 we may try to do this in a standardized way for all module.

Reply:
> Shouldn't you log a warning if more than one file is matched

That's something I thought about improving, although we're using fully qualified names combined with the tests folders which should result in unique files, so I'm not worried too much about this.

The reason why I'm using `inputFiles` instead of a single `inputFile` is to avoid catching exceptions if the file is not found, but actually, now I think that it would probably be more useful to do it that way and log an error if an expected file is not found.

Reply:
> BTW, this reminds me #64 we may try to do this in a standardised way for all module.

Agreed, I think we have been doing this in a wrong way using the `cwd` which leads to a whole load of path issues like, e.g. those discussed in #122. Using the FileSystem and PathResolver is the right way to go about this.

Reply:
Cool, so if everything goes right we may replace our old attempts with the new `resolve` helper method.
Probably work for another PR _(on which we should be sure to test all reported issues)_.